### PR TITLE
Merge Changes to testing and model outputs to main

### PR DESCRIPTION
**Pull request by luigirinaldi**
### New additions

- `hw_sim.py` file to contain various functions to emulate the hardware behaviour
- `Comparison` class in `test.py` to allow for different exit comparison functions to be evaluated at the same time, while tracking all desired statistics

### Changes

- CLI options to test over a linear range of thresholds and measure the performance
- CLI options to save the final layer output of the model as well as the values of the comparison functions used
- modified the `_forward_training` function in the models to output tensors to speed up inference when testing 

#### Data Analysis

All the code for generating various plots has been moved to a [dedicated branch](/biggsbenjamin/earlyexitnet/tree/data-analysis)

### CODE REVIEW

**Review by biggsbenjamin — CHANGES_REQUESTED** [review]
Happy to go through some of this tomorrow. Great work!

File: src/earlyexitnet/cli.py [hunk 1]
```diff
@@ -65,92 +83,145 @@ def test_only(args):
...
+    
+    final_object["test_vals"] = results
+    
+    save_path = f"{args.model_name}_{ts}.json"
+    
+    with open(save_path, 'a') as output:
+        output.write(json.dumps(final_object, indent=2))
+
+
+def run_test(datacoll,model,exits,top1_thr,entr_thr,loss_f,args, save_raw = False):
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/cli.py:147]: Tried to explain the difference as best I could, let me know if needs improvement

File: src/earlyexitnet/cli.py [hunk 2]
```diff
@@ -328,6 +413,12 @@ def main():
     #threshold inputs for testing, 1 or more args - user should know model
     parser.add_argument('-t1','--top1_threshold', nargs='+',type=float,required=False)
     parser.add_argument('-entr','--entr_threshold', nargs='+',type=float,required=False)
+    
+    parser.add_argument('-tr', '--threshold_range', nargs='+', type=float, required=False)
+    parser.add_argument('-ts', '--threshold_step', type=float, required=False)
+    
+    parser.add_argument('-sr', '--save_raw_softmax', type=bool, default=False, required=False,
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/cli.py:420]: Yes, it is some strange behaviour of the bool type arg parser, I changed it so it only check for the presence of the argument or accepts --no-save_raw as a False argument

File: src/earlyexitnet/testing_tools/test.py [hunk 3]
```diff
@@ -14,22 +17,138 @@
...
+                if isExit and not hasExited[batch]:
+                    self.exit_track.add_val(1, bin_index=exit_b)
+                    self.accu_track.update_correct(result_layer[batch], batched_correct_results[batch], bin_index=exit_b)
+                    hasExited[batch] = 1
+                    
+        stop = perf_counter()
+        
+        if self.test:
+            # concatenate along the batch dimension
+            self.raw_softmax = torch.cat((self.raw_softmax, sft_accum), dim=1) if self.raw_softmax is not None else sft_accum
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/testing_tools/test.py:104]: This is storing the values of the current softmax vector in a variable to be saved to file later

File: src/earlyexitnet/testing_tools/test.py [hunk 4]
```diff
@@ -14,22 +17,138 @@
...
         self.top1acc_thresholds = top1acc_thresholds
         self.entropy_thresholds = entropy_thresholds
+        
+        self.comp_funcs = comp_funcs
+        
+        self.save_raw = save_raw
+        
+        if self.save_raw:
+            self.true_result = None
+            self.raw_layer = None
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/testing_tools/test.py:150]: This sets up the variables to store the raw softmax vectors to be saved to file, the true result variable wasn't being used anymore

File: src/earlyexitnet/tools.py [hunk 5]
```diff
@@ -317,53 +319,31 @@ def _init_vars(self, #NOTE can you overloaded parent function
         self.set_length = set_length
         #init bins
         self.bin_num = bins
-        self.val_bins = np.zeros(bins,dtype=int)
+        self.val_bins = np.zeros(bins,dtype=np.float64)
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/tools.py:322]: Not sure, think this was changed while debugging something and else and was never corrected

File: src/earlyexitnet/tools.py [hunk 6]
```diff
@@ -317,53 +319,31 @@ def _init_vars(self, #NOTE can you overloaded parent function
...
+    
+        super().add_val(count,bin_index)
 
     def update_correct_list(self,res_list,lab_list=None): #list of lists of lists
         # [[bin0,bin1,...,binN],[bin0,bin1,...,binN],...,sampN], [label0,...labelN]
         if lab_list is not None:
-            assert len(res_list) == len(lab_list), "AccuTracker: sample size mismatch"
-            super().add_vals([[self.get_num_correct(res,label) for res in results]
-                                        for label,results in zip(lab_list,res_list)])
+            assert len(res_list[0]) == len(lab_list), "AccuTracker: sample size mismatch"
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/tools.py:343]: This is because the first dimension of the res_list is of size exit_nums, and the second dimension (the one indexed by [0]) is the size of the batch

File: src/earlyexitnet/testing_tools/test.py [hunk 7]
```diff
@@ -14,22 +17,138 @@
...
+                
+                # stack to separate exit dimension
+                sft_accum = torch.stack((sft_accum, raw_softmax),dim=0) if sft_accum is not None else raw_softmax
+                
+            else:
+                exit_result = self.compare_func(result_layer, self.exit_thresholds[exit_b])
+                        
+            # do this better by indexing into the arrays with the exit_result booleans 
+            # result_layer[exti_result] return a ternsor with only the layers which would've exited
+            for batch, isExit in enumerate(exit_result):
```
[the remark this answers is not in the document]
Comment (luigirinaldi) [reply src/earlyexitnet/testing_tools/test.py]: Yes, it infact yields approximately 10% speedup. also using masks to perform computation only on values that are yet to exit
